### PR TITLE
WIP: Supporting knative as a event processor for veba

### PR DESCRIPTION
**Pull request by rapuli**
## Summary

These changes allows VEBA to use knative as an Event processor. Knative Event processor will support channel-based 'Broker' implementation that uses channels for event routing and 'Trigger' which  subscribe  to events and filters from a specific Broker.

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [x] Please ensure that you are making a pull request against the **Development** branch
- [x] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [ ] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [ ] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commmits related to your change
- [ ] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [ ] Please include any relevant screenshots and/or output as part of your testing
- [ ] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe):

## Resolved Issues

List of Issues closed or resolved by this PR. Add multiple `Closes` keyword followed by the issue number (e.g. Closes #ISSUE-NUMBER)

* N/A

## Testing Verification

* UT is in progress. updating soon.

## Additional Information

* N/A
If you have any questions/comments, feel free to reach out to team on Slack [#vcenter-event-broker-appliance](https://vmwarecode.slack.com/archives/CQLT9B5AA)

Thank you from the VEBA Team! 🥳

### CODE REVIEW

File: vmware-event-router/deploy/event-router-config-knative.yaml
Comment (embano1) [inline vmware-event-router/deploy/event-router-config-knative.yaml:24]: Please sync the yaml file with your latest config

File: vmware-event-router/internal/config/v1alpha1/processor.go
Comment (embano1) [inline vmware-event-router/internal/config/v1alpha1/processor.go:43]: Why did you delete this line? Also your config does not have a comment?

File: vmware-event-router/internal/config/v1alpha1/processor.go
Comment (embano1) [inline vmware-event-router/internal/config/v1alpha1/processor.go:50]: As discussed, please provide an option for `insecureSSL` (TLS)

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:26]: This has changed recently (looks like you were developing from an older code base). Please export this struct and return `*KnativeProcessor` in the constructor.

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:29]: Needed?

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:30]: Needed?

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:13]: redundant import?

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:38]: Avoid empty lines here.

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:51]: Should this not be a negative test with error propagation?

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:78]: You access `cfg` above, so this `nil` check comes too late (panic).

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:91]: See comment above on empty lines

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:94]: Empty space needed, also remove "send events" since you're not sending here

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:97]: Redundant comment since this code is obvious.

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:95]: Please use an exponential algorithm.

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:99]: Remove empty line.

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:28]: unused?

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:106]: Can be simplified (and generally avoided) by returning early from the check above, e.g.

```go
if result := ...; cloudevents.IsACK(result) { ... return nil }
// continue with error path
```

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:109]: Don't use these fields, please see the documentation on the metrics and follow the guideline from the other processors.

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:113]: `ce.ID()` missing

**Review by embano1 — CHANGES_REQUESTED** [review]

**Comment by embano1** [comment]
I have ignored the tests for now since you're still working on them.
Please see the feedback. Also docs (README) changes are missing.

**Comment by embano1** [comment]
Also, please sign your commits to avoid the bot errors.

File: vmware-event-router/internal/config/v1alpha1/processor.go
Comment (embano1) [inline vmware-event-router/internal/config/v1alpha1/processor.go:32]: Please use same naming across the code base, e.g. `Knative`

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:66]: redundant log statements, just return the error

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:72]: see above

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:67]: should be: `return nil, errors.Wrap(err, "create protocol")`

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:73]: see above

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [inline vmware-event-router/internal/processor/knative.go:99]: `result` is of type `error` IIUC, so feel free to add this to log statements for better troubleshooting

File: vmware-event-router/internal/processor/knative.go
Comment (embano1) [reply vmware-event-router/internal/processor/knative.go:106]: Here I would print the error (`result`)

**Comment by embano1** [comment]
@rapuli again, PLEASE start signing your commits as mentioned above. The bot keeps spamming this PR and maintainer inboxes.

**Comment by embano1** [comment]
@rapuli you must sign all (previous) commits to unblock the bot or squash and merge them into one signed commit using `--force-with-lease` when pushing. Note (for the future) that this is not encouraged as it makes the review process harder to track changes.

**Comment by embano1** [comment]
Please also create an issue as described in the contributions/PR guidelines which this PR closes.

File: vmware-event-router/internal/config/v1alpha1/processor.go
Comment (rapuli) [reply vmware-event-router/internal/config/v1alpha1/processor.go:43]: Added Back the comment into the code.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go]: Removed retryInterval from the struct.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go]: Added code to print more logs, by using this verbose flag.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go]: Added error print.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go:25]: Updated the Code.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go:58]: Updated code as per the comments.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go:59]: Updated code as per the comment.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go:63]: Updated code as per the comment.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go:64]: Updated code as per the comment.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go:78]: removed comment // Send that Event.

File: vmware-event-router/internal/processor/knative.go
Comment (rapuli) [reply vmware-event-router/internal/processor/knative.go:88]: I could see this in openfaas
`of.lock.Lock()`
`of.stats.Invocations[res.Topic]++`
`of.lock.Unlock()
`
Can you tell me , which document are you referring for the metrics.

**Comment by embano1** [comment]
@rapuli not sure what you did, but looks like the PR is messed up with commits from other branches.

**Comment by embano1** [comment]
Dup, closing.

**Pull request closed without merging**
